parquet-variant-compute/src/unshred_variant.rs

```diff
@@ -0,0 +1,520 @@
+struct StructUnshredVariantBuilder<'a> {
+    value: Option<&'a arrow::array::BinaryViewArray>,
+    typed_value: &'a arrow::array::StructArray,
+    field_unshredders: IndexMap<&'a str, Option<UnshredVariantRowBuilder<'a>>>,
+}
+
+impl<'a> StructUnshredVariantBuilder<'a> {
+    fn try_new(value: Option<&'a BinaryViewArray>, typed_value: &'a StructArray) -> Result<Self> {
+        // Create unshredders for each field in constructor
+        let mut field_unshredders = IndexMap::new();
+        for (field, field_array) in typed_value.fields().iter().zip(typed_value.columns()) {
+            // Factory returns None for None/None case -- these are missing fields we should skip
+            let Some(field_array) = field_array.as_struct_opt() else {
+                return Err(ArrowError::InvalidArgumentError(format!(
+                    "Invalid shredded variant object field: expected Struct, got {}",
+                    field_array.data_type()
+                )));
+            };
+            let field_unshredder = make_unshred_variant_row_builder(field_array.try_into()?)?;
+            field_unshredders.insert(field.name().as_ref(), field_unshredder);
+        }
+
+        Ok(Self {
+            value,
+            typed_value,
+            field_unshredders,
+        })
+    }
+
+    fn append_row(
+        &mut self,
+        builder: &mut impl VariantBuilderExt,
+        metadata: &VariantMetadata,
+        index: usize,
+    ) -> Result<()> {
+        let value = handle_unshredded_case!(self, builder, metadata, index, true);
+
+        // If we get here, typed_value is valid and value may or may not be valid
+        let mut object_builder = builder.try_new_object()?;
+
+        // Process typed fields (skip empty builders that indicate missing fields)
+        for (field_name, field_unshredder_opt) in &mut self.field_unshredders {
+            if let Some(field_unshredder) = field_unshredder_opt {
+                let mut field_builder = ObjectFieldBuilder::new(field_name, &mut object_builder);
+                field_unshredder.append_row(&mut field_builder, metadata, index)?;
+            }
+        }
+
+        // Process any unshredded fields (partial shredding)
+        if let Some(value) = value {
+            let Variant::Object(object) = value else {
+                return Err(ArrowError::InvalidArgumentError(
+                    "Expected object in value field for partially shredded struct".to_string(),
+                ));
+            };
+
+            for (field_name, field_value) in object.iter() {
+                if self.field_unshredders.contains_key(field_name) {
+                    return Err(ArrowError::InvalidArgumentError(format!(
+                        "Field '{field_name}' appears in both typed_value and value",
+                    )));
+                }
+                object_builder.insert_bytes(field_name, field_value);
+            }
+        }
+
+        object_builder.finish();
+        Ok(())
+    }
+}
```

Comment:
I was wondering if we should have any tests for this kernel directly

I do realize the tests in parquet/tests/variant_integration.rs cover it pretty thoroughly

Maybe we can leave a comment to that effect

```suggestion
}
// Note: code is covered by tests in `parquet/tests/variant_integration.rs`
```